diff --git a/example/appdemo/maintask.c b/example/appdemo/maintask.c
index 461452b..1fe2380 100644
--- a/example/appdemo/maintask.c
+++ b/example/appdemo/maintask.c
@@ -22,6 +22,7 @@ extern void board_stduart_init(void);
 extern void board_dma_init(void);
 extern void board_gpio_init(void);
 extern void board_wifi_init(void);
+extern void board_kinit_init(kinit_t* init_args);
 
 /*  For user config
     kinit.argc = 0;
@@ -47,6 +48,7 @@ void aos_maintask(void* arg)
 {
     board_init();
     board_wifi_init();
+    board_kinit_init(&kinit);
     aos_components_init(&kinit);
 
 #ifndef AOS_BINS
diff --git a/example/basicfunction_app/aos_api_app/maintask.c b/example/basicfunction_app/aos_api_app/maintask.c
index b4ea0cc..d4b0fa6 100644
--- a/example/basicfunction_app/aos_api_app/maintask.c
+++ b/example/basicfunction_app/aos_api_app/maintask.c
@@ -22,6 +22,7 @@ extern void board_stduart_init(void);
 extern void board_dma_init(void);
 extern void board_gpio_init(void);
 extern void board_wifi_init(void);
+extern int board_kinit_init(kinit_t* init_args);
 
 /*  For user config
     kinit.argc = 0;
@@ -44,6 +45,7 @@ void board_init(void)
 void aos_maintask(void* arg)
 {
     board_init();
+	board_kinit_init(&kinit);
     aos_components_init(&kinit);
 
 #ifndef AOS_BINS
diff --git a/example/basicfunction_app/krhino_api_app/maintask.c b/example/basicfunction_app/krhino_api_app/maintask.c
index 28fb2d9..a7ae3d1 100644
--- a/example/basicfunction_app/krhino_api_app/maintask.c
+++ b/example/basicfunction_app/krhino_api_app/maintask.c
@@ -19,6 +19,7 @@ Set the correspondence in file platform\board\aaboard_demo\ucube.py.
 */
 extern void board_tick_init(void);
 extern void board_stduart_init(void);
+extern int board_kinit_init(kinit_t* init_args);
 
 /*  For user config
     kinit.argc = 0;
@@ -41,6 +42,7 @@ void board_init(void)
 void aos_maintask(void* arg)
 {
     board_init();
+	board_kinit_init(&kinit);
     aos_components_init(&kinit);
 
 #ifndef AOS_BINS
diff --git a/example/halappdemo/maintask.c b/example/halappdemo/maintask.c
index f28d72d..9f84140 100644
--- a/example/halappdemo/maintask.c
+++ b/example/halappdemo/maintask.c
@@ -19,6 +19,7 @@ Set the correspondence in file platform\board\aaboard_demo\ucube.py.
 */
 extern void board_tick_init(void);
 extern void board_stduart_init(void);
+extern void board_kinit_init(kinit_t* init_args);
 
 /*  For user config
     kinit.argc = 0;
@@ -43,6 +44,7 @@ void board_init(void)
 void aos_maintask(void* arg)
 {
     board_init();
+    board_kinit_init(&kinit);
     aos_components_init(&kinit);
 
 #ifndef AOS_BINS
diff --git a/example/linkkitappdemo/maintask.c b/example/linkkitappdemo/maintask.c
index 2ef8856..3c8f4d4 100644
--- a/example/linkkitappdemo/maintask.c
+++ b/example/linkkitappdemo/maintask.c
@@ -22,6 +22,7 @@ extern void board_stduart_init(void);
 extern void board_dma_init(void);
 extern void board_gpio_init(void);
 extern void board_network_init(void);
+extern void board_kinit_init(kinit_t* init_args);
 
 /*  For user config
     kinit.argc = 0;
@@ -50,6 +51,7 @@ void aos_maintask(void* arg)
 {
     board_init();
     board_network_init();
+    board_kinit_init(&kinit);
     aos_components_init(&kinit);
 
 #ifndef AOS_BINS
diff --git a/example/otaappdemo/maintask.c b/example/otaappdemo/maintask.c
index 2ef8856..de14f4a 100644
--- a/example/otaappdemo/maintask.c
+++ b/example/otaappdemo/maintask.c
@@ -22,6 +22,7 @@ extern void board_stduart_init(void);
 extern void board_dma_init(void);
 extern void board_gpio_init(void);
 extern void board_network_init(void);
+extern int board_kinit_init(kinit_t* init_args);
 
 /*  For user config
     kinit.argc = 0;
