diff -ruN zlib-1.2.11.orig/make-cross-x64.sh zlib-1.2.11/make-cross-x64.sh
--- zlib-1.2.11.orig/make-cross-x64.sh	1969-12-31 16:00:00.000000000 -0800
+++ zlib-1.2.11/make-cross-x64.sh	2017-10-21 21:09:10.000000000 -0700
@@ -0,0 +1,14 @@
+#!/bin/sh
+
+TARGET=x86_64-w64-mingw32
+INSTALL_PATH=/usr/local/$TARGET
+export PATH=$PATH:/usr/local/bin
+
+make -f win32/Makefile.gcc \
+    SHARED_MODE=1 \
+    LDFLAGS=-static-libgcc \
+    PREFIX=$TARGET- \
+    prefix=$INSTALL_PATH \
+    BINARY_PATH=$INSTALL_PATH/bin \
+    INCLUDE_PATH=$INSTALL_PATH/include \
+    LIBRARY_PATH=$INSTALL_PATH/lib $*
diff -ruN zlib-1.2.11.orig/make-cross-x86.sh zlib-1.2.11/make-cross-x86.sh
--- zlib-1.2.11.orig/make-cross-x86.sh	1969-12-31 16:00:00.000000000 -0800
+++ zlib-1.2.11/make-cross-x86.sh	2017-10-21 21:09:10.000000000 -0700
@@ -0,0 +1,14 @@
+#!/bin/sh
+
+TARGET=i686-w64-mingw32
+INSTALL_PATH=/usr/local/$TARGET
+export PATH=$PATH:/usr/local/bin
+
+make -f win32/Makefile.gcc \
+    SHARED_MODE=1 \
+    LDFLAGS=-static-libgcc \
+    PREFIX=$TARGET- \
+    prefix=$INSTALL_PATH \
+    BINARY_PATH=$INSTALL_PATH/bin \
+    INCLUDE_PATH=$INSTALL_PATH/include \
+    LIBRARY_PATH=$INSTALL_PATH/lib $*
